docs/docs/higher-order-components.md

```diff
@@ -0,0 +1,491 @@
+
+```js
+function onlyUpdateForKeys(InputComponent, keys) {
+  // Mutates the input component's prototype
+  // *Don't do this!*
+  InputComponent.prototype.shouldComponentUpdate = function(nextProps) {
+    for (let i = 0; i < keys.length; i++) {
+      let key = keys[i];
+      if (
+        !nextProps.hasOwnProperty(key) ||
+        !this.props.hasOwnProperty(key) ||
+        nextProps[key] !== this.props[key]
+      ) {
+        return false;
+      }
+    }
+    return true;
+  }
+  // The fact that we're returning the original input is a hint that it has
+  // been mutated.
+  return InputComponent;
+}
+
+// EnhancedComponent will only update when the `foo` or `bar` props change
+const EnhancedComponent = onlyUpdateForKeys(InputComponent, ['foo', 'bar'])
+```
+
+There are a few problems with this. One is that the input component cannot be reused separately from the enhanced component. More crucially, if you apply another HOC to `EnhancedComponent` that *also* mutates `shouldComponentUpdate`, the first HOC's functionality will be overridden! This HOC also won't work with function components, which do not have lifecycle methods.
+
+Mutating HOCs are a leaky abstraction—the consumer must know how they are implemented in order to avoid conflicts with other HOCs.
+
+Instead of mutation, HOCs should use composition, by wrapping the input component in a container component:
+
+```js
+function onlyUpdateForKeys(WrappedComponent, keys) {
+  return class extends React.Component {
+    shouldComponentUpdate(nextProps) {
+      for (let i = 0; i < keys.length; i++) {
+        let key = keys[i];
+        if (
+          !nextProps.hasOwnProperty(key) ||
+          !this.props.hasOwnProperty(key) ||
+          nextProps[key] !== this.props[key]
+        ) {
+          return false;
+        }
+      }
+      return true;
+    }
+    render() {
+      // Wraps the input component in a container, without mutating it. Good!
+      return <WrappedComponent {...this.props} />;
+    }
+  }
+}
+```
+
+This HOC has the same functionality as the mutating version while avoiding the potential for clashes. It works equally well with class and functional components. And because it's a pure function, it's composable with other HOCs, or even with itself:
+
+```js
+// EnhancedComponent will only update when the `foo`, `bar`, or `baz` props change
+let EnhancedComponent = onlyUpdateForKeys(InputComponent, ['foo', 'bar']);
+EnhancedComponent = onlyUpdateForKeys(EnhancedComponent, ['baz']);
+```
+
+You may have noticed similarities between HOCs and a pattern called **container components**. Container components are part of a strategy of separating responsibility between high-level and low-level concerns. Containers manage things like subscriptions and state, and pass props to components that handle things like rendering UI. HOCs use containers as part of their implementation. You can think of HOCs as paramaterized container component definitions.
```

Comment:
Oh I see what you mean. That's still a container to me, though I can see how this paragraph makes it seem like it's not a container unless it uses state or lifecycle methods.